python-checks/src/main/java/org/sonar/python/checks/IgnoredPureOperationsCheck.java

```diff
@@ -0,0 +1,389 @@
+    "frozenset.union",
+    "dict.copy",
+    "dict.dicfromkeys",
+    "dict.get",
+    "dict.items",
+    "dict.keys",
+    "dict.values"
+  ));
+
+  private static final Set<String> PURE_GETITEM_TYPES = new HashSet<>(Arrays.asList(
+    "str",
+    "bytes",
+    "bytearray",
+    "memoryview",
+    "list",
+    "tuple",
+    "range",
+    "dict"
+  ));
+
+  private static final Set<String> PURE_CONTAINS_TYPES = new HashSet<>();
+  static {
+    PURE_CONTAINS_TYPES.addAll(PURE_GETITEM_TYPES);
+    PURE_CONTAINS_TYPES.addAll(Arrays.asList(
+      "set",
+      "frozenset"
+    ));
+  }
+
+  private static class IgnoredPureOperation {
+    private Tree location;
+    private String functionName;
+
+    public IgnoredPureOperation(Tree location, String functionName) {
+      this.location = location;
+      this.functionName = functionName;
+    }
+  }
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.STATEMENT_LIST, ctx -> {
+      StatementList statementList = (StatementList) ctx.syntaxNode();
+
+      int numIssueStatements = 0;
+      List<IgnoredPureOperation> issueExpressions = new ArrayList<>();
+
+      for (Statement statement : statementList.statements()) {
+        if (statement.is(Tree.Kind.EXPRESSION_STMT)) {
+          ExpressionStatement expressionStatement = (ExpressionStatement) statement;
+          List<IgnoredPureOperation> issuesInExpressionStmt = getStatementIssues(expressionStatement);
+          issueExpressions.addAll(issuesInExpressionStmt);
+          if (!issuesInExpressionStmt.isEmpty()) {
+            numIssueStatements += 1;
+          }
+        }
+      }
+
+      if (!isExceptionalStatement(statementList, numIssueStatements)) {
+        issueExpressions.forEach(issue -> ctx.addIssue(issue.location, String.format(MESSAGE_FORMAT, issue.functionName)));
+      }
+    });
+  }
+
+  private static List<IgnoredPureOperation> getStatementIssues(ExpressionStatement expressionStatement) {
+    return expressionStatement.expressions().stream()
+      .map(IgnoredPureOperationsCheck::checkExpression)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
+  private static IgnoredPureOperation checkExpression(Expression expression) {
+    if (expression.is(Tree.Kind.CALL_EXPR)) {
+      CallExpression callExpression = (CallExpression) expression;
+      Symbol symbol = callExpression.calleeSymbol();
+      if (symbol == null) {
+        return null;
+      }
+
+      String fqn = symbol.fullyQualifiedName();
+      if (fqn != null && PURE_FUNCTIONS.contains(fqn)) {
+        return new IgnoredPureOperation(callExpression.callee(), fqn);
+      }
+    } else if (expression.is(Tree.Kind.SUBSCRIPTION)) {
+      SubscriptionExpression subscriptionExpression = ((SubscriptionExpression) expression);
+      InferredType type = subscriptionExpression.object().type();
+      if (PURE_GETITEM_TYPES.stream().anyMatch(type::canBeOrExtend)) {
```

Comment:
I think you'd want to use `type::canOnlyBe` here. Using type::canBeOrExtend here means you can have FPs if you have an unresolved type hierarchy. In fact, the only reason the current implementation doesn't raise an FP for `defaultdict` is because the symbol is hardcoded in `SymbolUtils::externalModulesSymbols`.

Also from what I can see from `expected-issues`, we shouldn't raise on subclasses anyway (otherwise, implementing an `InferredType::isOrExtend` could have been an option).

Reply:
I have updated the check, now we raise a lot less issues which could have been arguably false-positives. Thanks for the feedback!